src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs

```diff
@@ -194,35 +193,48 @@ internal bool HasEnumeratorCancellationAttribute
         }
 
+#nullable enable
+
         private ConstantValue DefaultSyntaxValue
         {
             get
             {
-                if (_lazyDefaultSyntaxValue == ConstantValue.Unset)
+                if (state.NotePartComplete(CompletionPart.StartDefaultSyntaxValue))
                 {
                     var diagnostics = DiagnosticBag.GetInstance();
-
-                    if (Interlocked.CompareExchange(
+                    var previousValue = Interlocked.CompareExchange(
                         ref _lazyDefaultSyntaxValue,
                         MakeDefaultExpression(diagnostics, out var binder, out var parameterEqualsValue),
-                        ConstantValue.Unset) == ConstantValue.Unset)
-                    {
-                        if (binder is not null && parameterEqualsValue is not null && !_lazyDefaultSyntaxValue.IsBad)
-                        {
-                            NullableWalker.AnalyzeIfNeeded(binder, parameterEqualsValue, diagnostics);
-                        }
-                        NullableAnalyzeParameterDefaultValueFromAttributes(diagnostics);
+                        ConstantValue.Unset);
+                    Debug.Assert(previousValue == ConstantValue.Unset);
+
+                    var completedOnThisThread = state.NotePartComplete(CompletionPart.EndDefaultSyntaxValue);
+                    Debug.Assert(completedOnThisThread);
 
-                        // This is a race condition where the thread that loses
-                        // the compare exchange tries to access the diagnostics
-                        // before the thread which won the race finishes adding
-                        // them. https://github.com/dotnet/roslyn/issues/17243
-                        AddDeclarationDiagnostics(diagnostics);
+                    if (binder is not null && parameterEqualsValue is not null && !_lazyDefaultSyntaxValue.IsBad)
+                    {
+                        NullableWalker.AnalyzeIfNeeded(binder, parameterEqualsValue, diagnostics);
+                        VerifyParamDefaultValueMatchesAttributeIfAny(_lazyDefaultSyntaxValue, parameterEqualsValue.Value.Syntax, diagnostics);
                     }
 
+                    AddDeclarationDiagnostics(diagnostics);
                     diagnostics.Free();
+
+                    completedOnThisThread = state.NotePartComplete(CompletionPart.EndDefaultSyntaxValueDiagnostics);
+                    Debug.Assert(completedOnThisThread);
                 }
 
+#if DEBUG
+                // If we accidentally access DefaultSyntaxValue in a reentrant manner, and we use a cancellation token
+                // that never expires, then the failure mode will be to spin wait forever.
+                // For debug purposes let's instead use a token which expires after a modest amount of time
+                // to wait for the default syntax value to be available.
+                var cts = new CancellationTokenSource(TimeSpan.FromSeconds(5));
```

Comment:
We need to pick a much higher number here. It is more than reasonable for a thread to hang for 5 secnds in CI

Reply:
Looks like integration tests use 1 minute to wait for things to complete in some places. I'll go with that as a starting point https://github.com/dotnet/roslyn/blob/42629354bc746019e212df6a0a991aeb75d12282/src/VisualStudio/IntegrationTest/TestUtilities/Helper.cs#L20